src/AnalyticsEventReporter.ts

```diff
@@ -1,38 +1,78 @@
+import { name, version } from '../package.json';
+import { post } from './post';
+import merge from './merge';
+import { setupRequestUrl } from './setupRequestUrl';
 
 /** Represents an reporter is responsible for reporting analytics events. */
 export class AnalyticsEventReporter implements AnalyticsEventService {
-
     private config: AnalyticsConfig;
     private payload?: EventPayload;
-
     /**
      * @param config - necessary analytics config: Must provide one and only
      * one of API Key or Bearer Token.
-     * 
+     *
      * @param payload - (optional) desired event values to report
      */
     constructor(config: AnalyticsConfig, payload?: EventPayload) {
-        const apiKeyIsSet = config.key !== undefined;
-        const bearerTokenIsSet = config.bearer !== undefined;
-        const configIsValid = (apiKeyIsSet || bearerTokenIsSet) && !(apiKeyIsSet && bearerTokenIsSet);
-        if (!configIsValid) {
-            throw new Error("Provide one and only one of API Key or Bearer Token.")
-        }
-        this.config = config;
-        this.payload = payload;
+      const apiKeyIsSet = config.key !== undefined;
+      const bearerTokenIsSet = config.bearer !== undefined;
+      const configIsValid = (apiKeyIsSet || bearerTokenIsSet) && !(apiKeyIsSet && bearerTokenIsSet);
+      if (!configIsValid) {
+        throw new Error('Provide one and only one of API Key or Bearer Token.');
+      }
+      this.config = config;
+      this.payload = payload;
     }
 
     with(payload: EventPayload): AnalyticsEventService {
-        const currentPayload = this.payload === undefined ? payload : merge(this.payload, payload);
-        return new AnalyticsEventReporter(this.config, currentPayload);
+      const currentPayload = this.payload === undefined ? payload : merge(this.payload, payload);
+      return new AnalyticsEventReporter(this.config, currentPayload);
     }
 
-    report(payload: object): Promise<EventAPIResponse> | boolean {
-        return false;
+    public async report(newPayload?: PartialPayload): Promise<boolean | EventAPIResponse> {
+      const finalPayload = this.payload && newPayload
+        ? merge(this.payload, newPayload)
+        : this.payload ?? newPayload ?? function(){throw Error('EventPayload is empty');}();
+
+      /** If session tracking is enabled, and sessionId was not already manually added to the event payload,
+       * call getOrSetupSessionId to set value.
+       */
+      if (this.config.sessionTrackingEnabled && !finalPayload.sessionId) {
+        finalPayload.sessionId = getOrSetupSessionId();
+      }
+
+      finalPayload.clientSdk
+        ? (finalPayload.clientSdk as Record<string, string>)[name] = version
+        : finalPayload.clientSdk = {
+          [name]: version,
+        };
+
+      finalPayload.authorization = this.config.key
+        ? 'KEY ' + this.config.key
+        : 'Bearer ' + this.config.bearer;
+
+      const res = await post(
```

Comment:
Thoughts on instead calling useBeacon() up here? And then based on that we can condition our response handling.

```
var useBeacon = useBeacon(...)

const res = await post(useBeacon,...)

if(useBeacon){
...
}
else {
// used fetch + keep-alive
}
```

we could even make two post methods `postWithBeacon` and `postWithFetch` so that its even less confusing in the handling